EventHubs migration guide (start) (#2)
richardpark-msft/azure-sdk-for-js: event hubs migration guide.md

# Migration Guide (EventHubs v2 to v5)

> ## Class changes

Reviewer: It's not incorrect to call me verbose, but I think a short summary of why there is a new library and why I would want to migrate would help establish context and set expectations.

Author: I added some intro text - mostly just to make sure they understood if this was the document for them (EH v2 users that just need some quick snippets for migration) or if they're new to the library (then I point them to the main package readme.md)

`EventHubClient` has been split into two clients - `EventHubProducerClient` (for sending
messages) and `EventHubConsumerClient` (for receiving messages).

`EventProcessorHost`'s functionality has been merged into `EventHubConsumerClient`.

> | In v2                                          | Equivalent in v5                                                 | Sample |

Reviewer: This may be more helpful and align better with samples if it were framed in scenarios rather than focused on a client.  Especially in cases where there may be multiple samples that would be relevant and helpful.

Author: I broke them apart a bit just to give them a little separation - so I have receiving, sending and construction as three separate little tables instead of one single table.

In V2, event handlers were passed as positional arguments to `receive`.

In V5, event handlers are passed as part of a `SubscriptionEventHandlers` shaped object.

For example, this code which receives from a partition in V2:

```typescript
const client = EventHubClient.createFromConnectionString(connectionString);
const rcvHandler = client.receive(partitionId, onMessageHandler, onErrorHandler, {
  eventPosition: EventPosition.fromStart(),
  consumerGroup: "$Default"
});
await rcvHandler.stop();
```

Becomes this in V5:

```typescript
const eventHubConsumerClient = new EventHubConsumerClient("$Default", connectionString);

const subscription = eventHubConsumerClient.subscribe(
  partitionId, {
    processInitialize: (initContext) => {
      initContext.setStartingPosition(EventPosition.fromStart());
    },
    processEvents: onMessageHandler,
    processError: onErrorHandler
});
  
await subscription.close();
```

See [`receiveEvents.ts`](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/receiveEvents.ts) 
for a sample program demonstrating this.

### Migrating code from `EventHubClient` to `EventHubProducerClient` for sending events

> In V2, there were multiple options on how to send data.

Reviewer: It may be helpful for those without much experience to give some of the core options a passing mention.   _(unless the goal is to not draw attention for comparison)_

Author: @chradek had a similar point. 

My goal is to get them to the next stage of "oh, this is the call I need to use" without bogging down too much in each one. 

Do you have a list of ones you consider typical options?

Reviewer: I was thinking the most likely to have been used:

- Send a single event using defaults
- Send an iterable of events using defaults

Author: We have less variety than you in JS ;)

We only have the sendBatch() call at this point and it's covered by the inline sample we have under "Migrating code from `EventHubClient` to `EventHubProducerClient` for sending events"

So they're all a single case.

  consumerGroup: consumerGroupName
});
await rcvHandler.stop();
```

Becomes this in V5:

```typescript
const eventHubConsumerClient = new EventHubConsumerClient(consumerGroupName, connectionString);

const subscription = eventHubConsumerClient.subscribe(
  partitionId, {
    processInitialize: (initContext) => {
      initContext.setStartingPosition(EventPosition.fromStart());
    },
    processEvents: onMessageHandler,
    processError: onErrorHandler
});
  
await subscription.close();
```

See [`receiveEvents.ts`](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/receiveEvents.ts) 
for a sample program demonstrating this.

### Migrating code from `EventHubClient` to `EventHubProducerClient` for sending events

In V2, there were multiple options on how to send data.

In V5, this has been consolidated into a more efficient `sendBatch` method. 
Batching merges information from multiple messages into a single send, reducing
the amount of network communication needed vs sending messages one at a time.

So in V2:
```typescript
const eventsToSend = [
  // events go here
];

const client = EventHubClient.createFromConnectionString(connectionString);

> console.log(`Sending event: ${eventData.body}`);

Reviewer 2: `eventData` doesnt exist anymore

Author: Oops, no it does not. Also, was calling the wrong send API.

// Would fail if the total size of events exceed the max size supported by the library.
await client.send(eventsToSend, partitionId);
```

In V5:
```typescript
const producer = new EventHubProducerClient(connectionString);

const eventsToSend = [
  // events go here
];

let batch = await producer.createBatch();

for (const event of eventsToSend) {
  // messages can fail to be added to the batch if they exceed the maximum size configured for
  // the EventHub.

>   const isAdded = batch.tryAdd({ body: event });

Reviewer 2: I would expect `eventsToSend` to already have the event data objects. So

```suggestion
  const isAdded = batch.tryAdd(event);
```

  
  if (!isAdded) {
    if (batch.count === 0) {
      // if we can't add it and the batch is empty that means the message we're trying to send
      // is too large, even when it would be the _only_ message in the batch.
      //

>       // To fix this you'll need to split the message up across multiple batches or 

Reviewer 2: How would they split the message? Not sure if this is our recommendation though...
How about throwing an error?

Author: I changed the wording - splitting implies we can help them in some way but we really can't. I did mention they need to find a way to shrink the message or just skip it, hopefully making it clear that their future is in their hands.

      // skip it. In this example, we'll skip the message.
      console.log(`Message was too large and can't be sent until it's made smaller. Skipping...`);
      continue;
    }

    // otherwise this just signals a good spot to send our batch
    console.log(`Batch is full - sending ${batch.count} messages as a single batch.`)
    await producer.sendBatch(batch);

    // and create a new one to house the next set of messages
    batch = await producer.createBatch();

>     continue;

Reviewer 2: In this model, wont the last event that resulted in `isAdded` being `false` getting skipped?

Author: Only with a binary-based computer.

(fixed. Also, fixed the sample it was based on)

  }
}

// send any remaining messages, if any.
if (batch.count > 0) {
  console.log(`Sending remaining ${batch.count} messages as a single batch.`)
  await producer.sendBatch(batch);
}
```

### Migrating code from `EventProcessorHost` to `EventHubConsumerClient` for receiving events

> In V2, `EventProcessorHost` allowed you to start receiving events, delivered via callbacks.

Reviewer 2: The selling point of EPH was the load balancing and checkpointing. We should call that out here

Author: This is less about what it's for (which they're familiar with) and more about how they achieve the same thing in the new library, right?

Reviewer 2: Yes, but if we are anyway talking about what EPH "allowed you to do", I'd prefer to have the full picture there if its doable. Something like

```suggestion
In V2, `EventProcessorHost` allowed you to balance the load between multiple instances of your program when receiving events.
```

Reviewer 2: Yes, but if we are anyway talking about what EPH "allowed you to do", I'd prefer to have the full picture there if its doable. Something like

```suggestion
In V2, `EventProcessorHost` allowed you to balance the load between multiple instances of your program when receiving events.
```

Reviewer 2: Yes, but if we are anyway talking about what EPH "allowed you to do", I'd prefer to have the full picture there if its doable. Something like

```suggestion
In V2, `EventProcessorHost` allowed you to balance the load between multiple instances of your program when receiving events.
```

  EventProcessorHost.createHostName(ephName),
  storageConnectionString,
  storageContainerName,
  ehConnectionString,
  {
    eventHubPath: eventHubName,
    onEphError: (error: any) => {
      console.log("[%s] Error: %O", ephName, error);
    }
  }
);

const onMessage: OnReceivedMessage = async (context: PartitionContext, event: EventData) => { }

const onError: OnReceivedError = (error: any) => {
  console.log("[%s] Received Error: %O", ephName, error);
};

await eph.start(onMessage, onError);
```

And in V5:
```typescript
import { EventHubConsumerClient, CheckpointStore } from "@azure/event-hubs";
import { ContainerClient } from "@azure/storage-blob";
import { BlobCheckpointStore } from "@azure/eventhubs-checkpointstore-blob";

const containerClient = new ContainerClient(storageConnectionString, storageContainerName);
const checkpointStore : CheckpointStore = new BlobCheckpointStore(containerClient);
const eventHubConsumerClient = new EventHubConsumerClient(consumerGroupName, ehConnectionString, eventHubName);

const subscription = eventHubConsumerClient.subscribe(
  partitionId, {
    // In V5 we deliver messages in batches, rather than a single message 

>     // at a time.

Reviewer 2: ```suggestion
    // at a time. You can control the batch size via the options passed to the client.
```

# Migration Guide (EventHubs v2 to v5)

> This document is intended for users that are familiar with EventHubs V2 (`@azure/event-hubs@2.x.x`) and wish 

Reviewer 2: ```suggestion
This document is intended for users that are familiar with EventHubs V2 (`@azure/event-hubs@2.x.x` & `@azure/event-processor-host@2.x.x`) and wish 
```

to migrate their application to EventHubs V5.

For a more general introduction to Event Hubs, see the [readme.md](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/README.md) 
for the @azure/event-hubs package.

## General changes

In the interest of simplifying the API surface we've made two distinct
clients, rather than having a single `EventHubClient`:
* [EventHubProducerClient](https://docs.microsoft.com/en-us/javascript/api/@azure/event-hubs/eventhubproducerclient?view=azure-node-preview)
  for sending messages.
* [EventHubConsumerClient](https://docs.microsoft.com/en-us/javascript/api/@azure/event-hubs/eventhubconsumerclient?view=azure-node-preview) 
  for receiving messages.

We've also merged the functionality from `EventProcessorHost` into 
`EventHubConsumerClient`, allowing `EventHubConsumerClient` to be the single

> point of entry for receiving of any type within Event Hubs.

Reviewer 2: ```suggestion
point of entry for receiving of any type (from single partition, all partitions, or with load balancing and checkpointing features) within Event Hubs.
```

> ### Construction

Reviewer 2: ```suggestion
### Client constructors
```

| In v2                                          | Equivalent in v5                                                 | Sample |
|------------------------------------------------|------------------------------------------------------------------|--------|

> | `EventHubClient.createFromConnectionString`    | `new EventHubProducerClient()` or `new EventHubConsumerClient()` | [receiveEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/receiveEvents.ts),  [sendEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/sendEvents.ts) |

Reviewer 2: ```suggestion
| `EventHubClient.createFromConnectionString()`    | `new EventHubProducerClient()` or `new EventHubConsumerClient()` | [receiveEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/receiveEvents.ts),  [sendEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/sendEvents.ts) |
```

> | `EventHubClient.createFromAadTokenCredentials` | `new EventHubProducerClient()` or `new EventHubConsumerClient()` | [usingAadAuth](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/usingAadAuth.ts)

Reviewer 2: ```suggestion
| `EventHubClient.createFromAadTokenCredentials()` | `new EventHubProducerClient()` or `new EventHubConsumerClient()` | [usingAadAuth](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/usingAadAuth.ts)
```

> ### Receiving events (subscribe)

Reviewer 2: I wouldnt suggest referring to "subscribe" in the header. If needed, you can add a one liner before the table about the fact that we now have a new name

```suggestion
### Receiving events 
```

# Migration Guide (EventHubs 2 to 5)

## Class changes

`EventHubClient` has been split into two clients - `EventHubProducerClient` (for sending

> messages) and `EventHubConsumerClient` (for subscribing to messages).

Reviewer 2: `for subscribing to messages` -> `for receiving messages` to have a symmetry between send and receive?
I understand that the method name is `subscribe`, but while reading this line the customer is probably not yet familiar with the method names and might wonder what "subscribe" means?

Author: Makes sense. It makes so much sense that that was what I wrote in my first draft and then I was like "well, the new method is called subscribe" ;)

| In v2                                       | Equivalent in v5                                                 | Sample |
|---------------------------------------------|------------------------------------------------------------------|--------|

> | `EventHubClient.createFromConnectionString` | `new EventHubProducerClient()` or `new EventHubConsumerClient()` | [All](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/)

Reviewer 2: What about when using AAD?

`EventHubClient.createFromAADTokenCredentials`  would become
`new EventHubProducerClient()` and `new EventHubConsumerClient()` using any of the [TokenCredentials](https://www.npmjs.com/package/@azure/identity)

Author: Awesome, will add.

| `EventHubClient.receive`                    | `EventHubConsumerClient.subscribe`                               | [receiveEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/receiveEvents.ts) |

> | `EventHubClient.receiveBatch`               | Removed in favor of `EventHubConsumerClient.subscribe`           | [receiveEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/receiveEvents.ts) |

Reviewer 2: > Removed in favor of `EventHubConsumerClient.subscribe` which now supports batching. Default size is 10 which you can change using the option `maxBatchSize`

Author: The only reason I didn't add this text was because it crowds the cell a bit and the sample does clearly show how to use it (with batching).

Just trying to make the table work as an "at a glance" thing.

> | `EventHubClient.createSender`               | `new EventHubProducerClient`                          | [sendEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/sendEvents.ts) |

Reviewer 2: I don't think `createSender()` existed in Track 1. Cant find it in the [reference docs](https://review.docs.microsoft.com/en-us/javascript/api/@azure/event-hubs/eventhubclient?view=azure-node-latest&branch=smoke-test-preview#methods)

Author: Nope, you're right. I'm not sure where I got that from.

> | `EventHubClient.createSender`               | `EventHubProducerClient.createProducer`                          | [sendEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/sendEvents.ts) |

Reviewer 3: ```suggestion
| `EventHubClient.createSender`               | `new EventHubProducerClient`                          | [sendEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/sendEvents.ts) |
```
We require instantiating the producer client. Do you know which parameters changed from v2 to v5 as well? I think we've made some required, some optional, and renamed others (epoch is now ownerLevel for instance)

Author: Might be good to identify which options we think of as key - quite a few things shifted around so I'd really like to stick to documenting the core scenarios if possible.

As a part of this though - if we decide that these are important parameters we should probably have a sample demonstrating their usage regardless.

> | `EventHubClient.send`                       | `EventHubConsumerClient.sendBatch`                               | [sendEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/sendEvents.ts) |

Reviewer 2: ```suggestion
| `EventHubClient.send`                       | `EventHubConsumerClient.sendBatch`                               | [sendEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/sendEvents.ts) which requires you to create a batch of events first.|
```

Author: You might be thinking of receiveEvents (not sendEvents.ts which does the sending).

There's a comment block in receiveEvents that walks you through running sendEvents to prime your EventHub.

Reviewer 2: My suggestion was more for the method `sendBatch` and not regarding the samples.

Current table contents tell the user that if they have been using `send`, they should now use `sendBatch`. But they also need to create a batch first, which they didnt need to do before. So, I believe its worthwhile calling that out.

> ### Migrating code from `EventHubClient` to `EventHubConsumerClient`

Reviewer 2: ```suggestion
### Migrating code from `EventHubClient` to `EventHubConsumerClient` for receiving events
```

In V2, event handlers were passed as positional arguments to `receive`.

In V5, event handlers are passed as part of a `SubscriptionEventHandlers` shaped object.

For example, this code which receives from a partition in V2:

> ```typescript

Reviewer 2: Add the line to create the client here as well to match with the v5 sample
```suggestion
```typescript
const client = EventHubClient.createFromConnectionString(connectionString);
```

const rcvHandler = client.receive(partitionId, onMessageHandler, onErrorHandler, {
  eventPosition: EventPosition.fromStart(),
  consumerGroup: "$Default"

> });

Reviewer 2: Add the line to stop the receiving of events to match with the v5 sample
```suggestion
});
await rcvHandler.stop();
```

```

Becomes this in V5:

```typescript
const eventHubConsumerClient = new EventHubConsumerClient("$Default", connectionString);

const subscription = eventHubConsumerClient.subscribe(
  partitionId, {
    processInitialize: (initContext) => {
      initContext.setStartingPosition(EventPosition.fromStart());
    },
    processEvents: onMessageHandler,
    processError: onErrorHandler
});
  
await subscription.close();
```

See [`receiveEvents.ts`](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/receiveEvents.ts) 
for a sample program demonstrating this.

> ### Migrating code from `EventHubClient` to `EventHubProducerClient`

Reviewer 2: ```suggestion
### Migrating code from `EventHubClient` to `EventHubProducerClient` for sending events.
```

In V2, there were multiple options on how to send data.

> In V5, this has been consolidated into a more efficient `sendBatch` method.

Reviewer 2: Add few words as to why send batch is more efficient?

Author: Done.

So in V2:
```typescript

> console.log(`Sending event: ${eventData.body}`);

Reviewer 2: Add the line to create the client here as well to match with the v5 sample

```suggestion
const client = EventHubClient.createFromConnectionString(connectionString);
console.log(`Sending event: ${eventData.body}`);
```

> await client.send(eventData, partitionId);

Reviewer 2: Since we want to highlight why sending with batches is better, should we consider the code snippet to reflect a function that takes an array of events? This is close to what we want to https://github.com/Azure/azure-sdk-for-js/issues/6254

The old sample would then be

```javascript
async function sendMyEvents(events) {
      await client.send(events); // Will fail if the total size of events exceed the max size supported by the library.
}
```

The newer one would be
```javascript
async function sendMyEvents(events) {
     const batch = await eventHubProducerClient.createBatch();
     let leftOverEvents = [];
     for (let i = 0; i < events.length; i++) {
        if (!batch.tryAdd(events[i]) {
               if (i === 0) {
                   // throw error here as we cannot batch the single event, too big
              }
               leftOverEvents = events.slice(i);
               break;
        }
     }
     await client.sendBatch(batch);
     await sendMyEvents(leftOverEvents);
}
```

# Migration Guide (EventHubs 2 to 5)

## Class changes

> `EventHubClient` has been split into two clients - `EventHubProducerClient` (for sending

Reviewer 2: We need to also talk about how the EPH class is not needed anymore and is replaced by EventHubConsumerClient

Author: Added a line (haven't done the sample yet)

Author: Looks like this is a duplicate comment to below. Resolving this one.

| `EventHubClient.createFromConnectionString`    | `new EventHubProducerClient()` or `new EventHubConsumerClient()` | [All](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/) |
| `EventHubClient.createFromAadTokenCredentials` | `new EventHubProducerClient()` or `new EventHubConsumerClient()` | [usingAadAuth](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/usingAadAuth.ts)
| `EventHubClient.receive`                       | `EventHubConsumerClient.subscribe`                               | [receiveEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/receiveEvents.ts) |
| `EventHubClient.receiveBatch`                  | Removed in favor of `EventHubConsumerClient.subscribe`           | [receiveEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/receiveEvents.ts) |
| `EventHubClient.send`                          | `EventHubConsumerClient.sendBatch`                               | [sendEvents](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/event-hubs/samples/sendEvents.ts) |

> ### Migrating code from `EventHubClient` to `EventHubConsumerClient` for receiving events

Reviewer 2: Need a section for 

> Migrating code from `EventProcessorHost` to `EventHubConsumerClient` for receiving events from all partitions

Author: Added - I think I captured the big differences:
* EPH used to have multiple error callbacks. Now there's one.
* EPH delivered a single message at a time. Now you get a batch of messages.
* Construction differences.

Let me know if there should be more - not as familiar with EPH.
